pkg/sbom/scanner/scanner.go

```diff
@@ -86,129 +121,189 @@ func (s *Scanner) enoughDiskSpace(opts sbom.ScanOptions) error {
 					}
 				}
-			case request, ok := <-s.scanQueue:
-				// Channel has been closed we should exit
-				if !ok {
-					break loop
-				}
-				telemetry.SBOMAttempts.Inc(request.Collector(), request.Type())
-
-				collector := request.collector
-				if err := s.enoughDiskSpace(request.collector.Options()); err != nil {
-					var imgMeta *workloadmeta.ContainerImageMetadata
-					if store := workloadmeta.GetGlobalStore(); store != nil {
-						img, err := store.GetImage(request.ID())
-						if err != nil {
-							imgMeta = img
-						}
-					}
-
-					result := sbom.ScanResult{
-						ImgMeta: imgMeta,
-						Error:   fmt.Errorf("failed to check current disk usage: %w", err),
-					}
-					request.sendResult(&result)
-					telemetry.SBOMFailures.Inc(request.Collector(), request.Type(), "disk_space")
-					continue
-				}
+				s.cacheMutex.Unlock()
+			}
+		}
+	}()
+}
 
-				scanTimeout := request.collector.Options().Timeout
-				if scanTimeout == 0 {
-					scanTimeout = defaultScanTimeout
-				}
+func (s *Scanner) start(ctx context.Context) {
+	if s.running {
+		return
+	}
+	s.running = true
+	s.startCacheCleaner(ctx)
+	s.startScanRequestHandler(ctx)
+}
 
-				scanContext, cancel := context.WithTimeout(ctx, scanTimeout)
-				createdAt := time.Now()
-				scanResult := collector.Scan(scanContext, request.ScanRequest)
-				generationDuration := time.Since(createdAt)
-				scanResult.CreatedAt = createdAt
-				scanResult.Duration = generationDuration
-				if scanResult.Error != nil {
-					telemetry.SBOMFailures.Inc(request.Collector(), request.Type(), "scan")
-				} else {
-					telemetry.SBOMGenerationDuration.Observe(generationDuration.Seconds(), request.Collector(), request.Type())
-				}
-				cancel()
-				request.sendResult(&scanResult)
-				if request.collector.Options().WaitAfter != 0 {
-					t := time.NewTimer(request.collector.Options().WaitAfter)
-					select {
-					case <-ctx.Done():
-					case <-t.C:
-					}
-					t.Stop()
-				}
+func (s *Scanner) startScanRequestHandler(ctx context.Context) {
+	go func() {
+		<-ctx.Done()
+		s.scanQueue.ShutDown()
+	}()
+	go func() {
+		for {
+			r, shutdown := s.scanQueue.Get()
+			if shutdown {
+				break
 			}
+			handleScanRequest(ctx, r, s)
 		}
-
 		for _, collector := range collectors.Collectors {
 			collector.Shutdown()
 		}
 	}()
 }
 
-// Start starts the scanner
-func (s *Scanner) Start(ctx context.Context) {
-	s.startOnce.Do(func() {
-		s.start(ctx)
-	})
+func handleScanRequest(ctx context.Context, r interface{}, s *Scanner) {
+	request, ok := r.(sbom.ScanRequest)
+	if !ok {
+		_ = log.Errorf("invalid scan request type '%T'", r)
+		s.scanQueue.Forget(r)
+		s.scanQueue.Done(r)
+		return
+	}
+	telemetry.SBOMAttempts.Inc(request.Collector(), request.Type())
+	collector, ok := collectors.Collectors[request.Collector()]
+	if !ok {
+		_ = log.Errorf("invalid collector '%s'", request.Collector())
+		s.scanQueue.Forget(request)
+		s.scanQueue.Done(request)
```

Comment:
can we move the `Done()` in a defer() {} 
and remove it from the `getImageMetadata()` function?
because for clarity this `s.scanQueue.Done(request)` should be call only in one place at the end of the `handleScanRequest()` execution